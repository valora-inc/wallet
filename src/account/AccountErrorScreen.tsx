import React from 'react'
import { StyleSheet, Text } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import TextButton from 'src/components/TextButton'
import fontStyles from 'src/styles/fonts'
import { Spacing } from 'src/styles/styles'

interface Props {
  title: string
  testID: string
  description: string
  onPress(): void
  buttonLabel: string
  onPressSecondary?(): void
  secondaryButtonLabel?: string
}

function AccountErrorScreen({
  title,
  testID,
  description,
  onPress,
  buttonLabel,
  secondaryButtonLabel,
  onPressSecondary,
}: Props) {
  return (
    <SafeAreaView style={styles.content}>
      <Text style={styles.title} testID={testID}>
        {title}
      </Text>
      <Text style={styles.body}>{description}</Text>
      <TextButton onPress={onPress} testID={`${testID}Button`}>
        {buttonLabel}
      </TextButton>
      {secondaryButtonLabel && onPressSecondary && (
        <TextButton
          style={styles.secondaryButton}
          onPress={onPressSecondary}
          testID={`${testID}ButtonSecondary`}
        >
          {secondaryButtonLabel}
        </TextButton>
      )}
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  content: {
    flex: 1,
    marginHorizontal: Spacing.Thick24,
    justifyContent: 'center',
    alignItems: 'center',
  },
  title: {
    ...fontStyles.h2,
    textAlign: 'center',
    paddingBottom: Spacing.Regular16,
    paddingTop: Spacing.Regular16,
  },
  body: {
    ...fontStyles.regular,
    textAlign: 'center',
    paddingBottom: Spacing.Thick24,
  },
  secondaryButton: {
    paddingTop: Spacing.Thick24,
  },
})

export default AccountErrorScreen
