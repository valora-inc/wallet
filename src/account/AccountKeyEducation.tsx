import { StackScreenProps, TransitionPresets } from '@react-navigation/stack'
import React, { useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import Education, { EducationTopic, EmbeddedNavBar } from 'src/account/Education'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { BtnTypes } from 'src/components/Button'
import { accountKey1, accountKey2, accountKey3, accountKey4 } from 'src/images/Images'
import { noHeader } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'

type Props = StackScreenProps<StackParamList, Screens.AccountKeyEducation>

export default function AccountKeyEducation(props: Props) {
  function onComplete() {
    ValoraAnalytics.track(OnboardingEvents.backup_education_complete)
    if (props.route.params?.nextScreen) {
      navigate(props.route.params?.nextScreen)
    } else {
      navigate(Screens.BackupPhrase)
    }
  }

  const { t } = useTranslation()

  const steps = useSteps()

  useEffect(() => {
    ValoraAnalytics.track(OnboardingEvents.backup_education_start)
  }, [])

  return (
    <Education
      embeddedNavBar={EmbeddedNavBar.Close}
      stepInfo={steps}
      onFinish={onComplete}
      finalButtonText={t('completeEducation')}
      buttonText={t('next')}
      finalButtonType={BtnTypes.PRIMARY}
    />
  )
}

AccountKeyEducation.navigationOptions = {
  ...noHeader,
  ...TransitionPresets.ModalTransition,
}

function useSteps() {
  const { t } = useTranslation()
  return React.useMemo(
    () =>
      [
        { image: accountKey1, topic: EducationTopic.backup },
        { image: accountKey2, topic: EducationTopic.backup },
        { image: accountKey3, topic: EducationTopic.backup },
        { image: accountKey4, topic: EducationTopic.backup },
      ].map((step, index) => {
        return {
          ...step,
          title: t(`guide.${index}.title`),
          text: t(`guide.${index}.text`),
        }
      }),
    []
  )
}
