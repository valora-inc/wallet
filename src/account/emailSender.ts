import { openComposer } from 'react-native-email-link'
import * as RNFS from 'react-native-fs'
import Mailer from 'react-native-mail'
import { CELO_SUPPORT_EMAIL_ADDRESS } from 'src/config'
import Logger from 'src/utils/Logger'

export interface Email {
  subject: string
  recipients: [string]
  body: string
  isHTML: boolean
  attachments?: Array<{
    path: string
    type: string
    name: string
  }>
}

export async function sendEmailWithNonNativeApp(
  emailSubect: string,
  message: string,
  deviceInfo?: {},
  combinedLogsPath?: string | false
) {
  try {
    const supportLogsMessage = combinedLogsPath
      ? `Support logs: ${!combinedLogsPath || (await RNFS.readFile(combinedLogsPath))}`
      : ''
    await openComposer({
      to: CELO_SUPPORT_EMAIL_ADDRESS,
      subject: emailSubect,
      body: `${message}\n${deviceInfo ? JSON.stringify(deviceInfo) : ''}\n${supportLogsMessage}`,
    })
    return { success: true }
  } catch (error) {
    return { error }
  }
}

export async function sendEmail(
  email: Email,
  deviceInfo?: {},
  combinedLogsPath: string | false = false
) {
  return new Promise<void>((resolve, reject) => {
    // Try to send with native mail app with logs as attachment
    // if fails user can choose mail app but logs sent in message
    Mailer.mail(email, async (error: any, event: string) => {
      if (event === 'sent') {
        resolve()
      } else if (error) {
        const emailSent = await sendEmailWithNonNativeApp(
          email.subject,
          email.body,
          deviceInfo,
          combinedLogsPath
        )
        if (emailSent.success) {
          resolve()
        } else {
          Logger.showError(error + ' ' + emailSent.error)
          reject()
        }
      }
    })
  })
}
