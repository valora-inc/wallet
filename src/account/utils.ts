import { parsePhoneNumber } from '@celo/phone-utils'

const ADDRESS_LENGTH = 42 // TODO(ACT-1173): see if this can be replaced with a viem helper

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}

export function getPhoneNumberDetails(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      // Show international display number to avoid confusion
      internationalPhoneNumber: phoneDetails.displayNumberInternational,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      internationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
