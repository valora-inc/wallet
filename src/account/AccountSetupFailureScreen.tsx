import React from 'react'
import { useTranslation } from 'react-i18next'
import RNExitApp from 'react-native-exit-app'
import AccountErrorScreen from 'src/account/AccountErrorScreen'
import { noHeaderGestureDisabled } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'

function AccounSetupFailureScreen() {
  const { t } = useTranslation()

  const onPressCloseApp = () => {
    RNExitApp.exitApp()
  }

  const onPressContactSupport = () => {
    navigate(Screens.SupportContact)
  }

  return (
    <AccountErrorScreen
      title={t('accountSetupFailed')}
      testID="AccountSetupFailure"
      description={t('accountSetupFailedDescription')}
      onPress={onPressCloseApp}
      buttonLabel={t('closeApp')}
      onPressSecondary={onPressContactSupport}
      secondaryButtonLabel={t('contactSupport')}
    />
  )
}

AccounSetupFailureScreen.navOptions = noHeaderGestureDisabled

export default AccounSetupFailureScreen
