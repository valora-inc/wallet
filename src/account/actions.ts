import { PincodeType } from 'src/account/reducer'

export enum Actions {
  CHOOSE_CREATE_ACCOUNT = 'ACCOUNT/CHOOSE_CREATE',
  CHOOSE_RESTORE_ACCOUNT = 'ACCOUNT/CHOOSE_RESTORE',
  START_STORE_WIPE_RECOVERY = 'ACCOUNT/START_STORE_WIPE_RECOVERY',
  CANCEL_CREATE_OR_RESTORE_ACCOUNT = 'ACCOUNT/CANCEL_CREATE_OR_RESTORE_ACCOUNT',
  SET_NAME = 'ACCOUNT/SET_NAME',
  SET_PHONE_NUMBER = 'ACCOUNT/SET_PHONE_NUMBER',
  SAVE_NAME = 'ACCOUNT/SAVE_NAME',
  DEV_MODE_TRIGGER_CLICKED = 'ACCOUNT/NAME_CLICKED',
  PHOTOSNUX_CLICKED = 'ACCOUNT/PHOTOSNUX_CLICKED',
  SET_PINCODE_SUCCESS = 'ACCOUNT/SET_PINCODE_SUCCESS',
  SET_PINCODE_FAILURE = 'ACCOUNT/SET_PINCODE_FAILURE',
  SET_ACCOUNT_CREATION_TIME = 'ACCOUNT/SET_ACCOUNT_CREATION_TIME',
  INITIALIZE_ACCOUNT = 'ACCOUNT/INITIALIZE_ACCOUNT',
  INITIALIZE_ACCOUNT_SUCCESS = 'ACCOUNT/INITIALIZE_ACCOUNT_SUCCESS',
  INITIALIZE_ACCOUNT_FAILURE = 'ACCOUNT/INITIALIZE_ACCOUNT_FAILURE',
  SET_BACKUP_COMPLETED = 'ACCOUNT/SET_BACKUP_COMPLETED',
  DISMISS_GET_VERIFIED = 'ACCOUNT/DISMISS_GET_VERIFIED',
  DISMISS_GOLD_EDUCATION = 'ACCOUNT/DISMISS_GOLD_EDUCATION',
  SET_USER_CONTACT_DETAILS = 'ACCOUNT/SET_USER_CONTACT_DETAILS',
  ACCEPT_TERMS = 'ACCOUNT/ACCEPT_TERMS',
  CLEAR_STORED_ACCOUNT = 'ACCOUNT/CLEAR_STORED_ACCOUNT',
  PROFILE_UPLOADED = 'ACCOUNT/PROFILE_UPLOADED',
  SAVE_SIGNED_MESSAGE = 'ACCOUNT/SAVE_SIGNED_MESSAGE',
  SET_CELO_EDUCATION_COMPLETED = 'ACCOUNT/SET_CELO_EDUCATION_COMPLETED',
  RECOVERY_PHRASE_IN_ONBOARDING_STARTED = 'ACCOUNT/RECOVERY_PHRASE_IN_ONBOARDING_STARTED',
  RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED = 'ACCOUNT/RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED',
}

export interface ChooseCreateAccountAction {
  type: Actions.CHOOSE_CREATE_ACCOUNT
  now: number
}
interface ChooseRestoreAccountAction {
  type: Actions.CHOOSE_RESTORE_ACCOUNT
}

interface StartStoreWipeRecoveryAction {
  type: Actions.START_STORE_WIPE_RECOVERY
  accountToRecover: string
}

interface CancelCreateOrRestoreAccountAction {
  type: Actions.CANCEL_CREATE_OR_RESTORE_ACCOUNT
}

interface SetNameAction {
  type: Actions.SET_NAME
  name: string
}

interface SetPhoneNumberAction {
  type: Actions.SET_PHONE_NUMBER
  e164PhoneNumber: string
  countryCode: string
}

interface SaveNameAction {
  type: Actions.SAVE_NAME
  name: string
}
interface DevModeTriggerClickedAction {
  type: Actions.DEV_MODE_TRIGGER_CLICKED
}

interface AcceptTermsAction {
  type: Actions.ACCEPT_TERMS
}

interface PhotosNUXClickedAction {
  type: Actions.PHOTOSNUX_CLICKED
}

interface SetPincodeSuccessAction {
  type: Actions.SET_PINCODE_SUCCESS
  pincodeType: PincodeType
}

interface SetPincodeFailureAction {
  type: Actions.SET_PINCODE_FAILURE
}

interface InitializeAccountAction {
  type: Actions.INITIALIZE_ACCOUNT
}

interface InitializeAccountSuccessAction {
  type: Actions.INITIALIZE_ACCOUNT_SUCCESS
}

interface SetAccountCreationAction {
  type: Actions.SET_ACCOUNT_CREATION_TIME
  now: number
}

interface SetBackupCompletedAction {
  type: Actions.SET_BACKUP_COMPLETED
}

interface DismissGetVerifiedAction {
  type: Actions.DISMISS_GET_VERIFIED
}

interface DismissGoldEducationAction {
  type: Actions.DISMISS_GOLD_EDUCATION
}

interface SetContactDetailsAction {
  type: Actions.SET_USER_CONTACT_DETAILS
  contactId: string
  thumbnailPath: string | null
}

export interface ClearStoredAccountAction {
  type: Actions.CLEAR_STORED_ACCOUNT
  account: string
}

interface ProfileUploadedAction {
  type: Actions.PROFILE_UPLOADED
}

interface SaveSignedMessage {
  type: Actions.SAVE_SIGNED_MESSAGE
}

interface SetCeloEducationCompletedAction {
  type: Actions.SET_CELO_EDUCATION_COMPLETED
  celoEducationCompleted: boolean
}

interface RecoveryPhraseInOnboardingStarted {
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_STARTED
}

interface RecoveryPhraseInOnboardingCompleted {
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED
}

export type ActionTypes =
  | ChooseCreateAccountAction
  | ChooseRestoreAccountAction
  | StartStoreWipeRecoveryAction
  | CancelCreateOrRestoreAccountAction
  | SetNameAction
  | SetPhoneNumberAction
  | SaveNameAction
  | DevModeTriggerClickedAction
  | PhotosNUXClickedAction
  | SetPincodeSuccessAction
  | SetPincodeFailureAction
  | InitializeAccountAction
  | InitializeAccountSuccessAction
  | SetAccountCreationAction
  | SetBackupCompletedAction
  | DismissGetVerifiedAction
  | DismissGoldEducationAction
  | SetContactDetailsAction
  | AcceptTermsAction
  | ClearStoredAccountAction
  | ProfileUploadedAction
  | SaveSignedMessage
  | SetCeloEducationCompletedAction
  | RecoveryPhraseInOnboardingStarted
  | RecoveryPhraseInOnboardingCompleted

export function chooseCreateAccount(now: number): ChooseCreateAccountAction {
  return {
    type: Actions.CHOOSE_CREATE_ACCOUNT,
    now,
  }
}

export function chooseRestoreAccount(): ChooseRestoreAccountAction {
  return {
    type: Actions.CHOOSE_RESTORE_ACCOUNT,
  }
}

export function startStoreWipeRecovery(accountToRecover: string): StartStoreWipeRecoveryAction {
  return {
    type: Actions.START_STORE_WIPE_RECOVERY,
    accountToRecover,
  }
}

export function cancelCreateOrRestoreAccount(): CancelCreateOrRestoreAccountAction {
  return {
    type: Actions.CANCEL_CREATE_OR_RESTORE_ACCOUNT,
  }
}

export function acceptTerms(): AcceptTermsAction {
  return {
    type: Actions.ACCEPT_TERMS,
  }
}

export function saveName(name: string): SaveNameAction {
  return {
    type: Actions.SAVE_NAME,
    name,
  }
}
export const devModeTriggerClicked = (): DevModeTriggerClickedAction => ({
  type: Actions.DEV_MODE_TRIGGER_CLICKED,
})

export const setPincodeSuccess = (pincodeType: PincodeType): SetPincodeSuccessAction => ({
  type: Actions.SET_PINCODE_SUCCESS,
  pincodeType,
})

export const initializeAccount = (): InitializeAccountAction => ({
  type: Actions.INITIALIZE_ACCOUNT,
})

export const initializeAccountSuccess = (): InitializeAccountSuccessAction => ({
  type: Actions.INITIALIZE_ACCOUNT_SUCCESS,
})

export const setAccountCreationTime = (now: number): SetAccountCreationAction => ({
  type: Actions.SET_ACCOUNT_CREATION_TIME,
  now,
})

export const setBackupCompleted = (): SetBackupCompletedAction => ({
  type: Actions.SET_BACKUP_COMPLETED,
})

export const dismissGetVerified = (): DismissGetVerifiedAction => ({
  type: Actions.DISMISS_GET_VERIFIED,
})

export const dismissGoldEducation = (): DismissGoldEducationAction => ({
  type: Actions.DISMISS_GOLD_EDUCATION,
})

export const setUserContactDetails = (
  contactId: string,
  thumbnailPath: string | null
): SetContactDetailsAction => ({
  type: Actions.SET_USER_CONTACT_DETAILS,
  contactId,
  thumbnailPath,
})

export const clearStoredAccount = (account: string): ClearStoredAccountAction => ({
  type: Actions.CLEAR_STORED_ACCOUNT,
  account,
})

export const saveSignedMessage = (): SaveSignedMessage => ({
  type: Actions.SAVE_SIGNED_MESSAGE,
})

export const setGoldEducationCompleted = (): SetCeloEducationCompletedAction => ({
  type: Actions.SET_CELO_EDUCATION_COMPLETED,
  celoEducationCompleted: true,
})

export const recoveryPhraseInOnboardingStarted = (): RecoveryPhraseInOnboardingStarted => ({
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_STARTED,
})

export const recoveryPhraseInOnboardingCompleted = (): RecoveryPhraseInOnboardingCompleted => ({
  type: Actions.RECOVERY_PHRASE_IN_ONBOARDING_COMPLETED,
})
