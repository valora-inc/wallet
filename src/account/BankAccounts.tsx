import { StackScreenProps } from '@react-navigation/stack'
import React, { useEffect, useLayoutEffect, useState } from 'react'
import { useAsync } from 'react-async-hook'
import { useTranslation } from 'react-i18next'
import {
  ActivityIndicator,
  BackHandler,
  Image,
  ScrollView,
  StyleSheet,
  Text,
  View,
} from 'react-native'
import { usePlaidEmitter } from 'react-native-plaid-link-sdk'
import { useDispatch, useSelector } from 'react-redux'
import { plaidParamsSelector } from 'src/account/selectors'
import { showError } from 'src/alert/actions'
import { CICOEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import BorderlessButton from 'src/components/BorderlessButton'
import OptionsChooser from 'src/components/OptionsChooser'
import BackChevron from 'src/icons/BackChevron'
import PlusIcon from 'src/icons/PlusIcon'
import TripleDotVertical from 'src/icons/TripleDotVertical'
import {
  BankAccount,
  deleteFinclusiveBankAccount,
  getFinclusiveBankAccounts,
  verifyWalletAddress,
} from 'src/in-house-liquidity'
import { emptyHeader } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { TopBarIconButton } from 'src/navigator/TopBarButton'
import { StackParamList } from 'src/navigator/types'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'
import variables from 'src/styles/variables'
import Logger from 'src/utils/Logger'
import { walletAddressSelector } from 'src/web3/selectors'
import openPlaid, { handleOnEvent } from './openPlaid'

type Props = StackScreenProps<StackParamList, Screens.BankAccounts>

const TAG = 'BankAccounts'
function BankAccounts({ navigation, route }: Props) {
  const { t } = useTranslation()
  const dispatch = useDispatch()
  usePlaidEmitter(handleOnEvent)
  const [isOptionsVisible, setIsOptionsVisible] = useState(false)
  const [selectedBankId, setSelectedBankId] = useState(0)
  const walletAddress = useSelector(walletAddressSelector)
  const plaidParams = useSelector(plaidParamsSelector)
  const { newPublicToken, fromSyncBankAccountScreen } = route.params

  const header = () => {
    return (
      <View style={styles.header}>
        <Text style={fontStyles.navigationHeader}>{t('bankAccountsScreen.header')}</Text>
      </View>
    )
  }

  const navigateToSettings = () => {
    navigate(Screens.Settings)
  }

  useEffect(() => {
    if (fromSyncBankAccountScreen === true) {
      // Prevent back button on Android when previous screen is SyncBankAccountScreen
      const backPressListener = () => true
      BackHandler.addEventListener('hardwareBackPress', backPressListener)
      return () => BackHandler.removeEventListener('hardwareBackPress', backPressListener)
    }
  }, [])

  useLayoutEffect(() => {
    navigation.setOptions({
      headerTitle: header,
      headerLeft: () => (
        <TopBarIconButton
          icon={<BackChevron />}
          onPress={navigateToSettings}
          style={styles.backButton}
          testID="backButton"
        />
      ),
      // Prevent swiping back on iOS when previous screen is SyncBankAccountScreen
      gestureEnabled: !fromSyncBankAccountScreen,
    })
  }, [navigation])

  const bankAccounts = useAsync(async () => {
    try {
      const accounts = await getFinclusiveBankAccounts(verifyWalletAddress({ walletAddress }))
      return accounts
    } catch (error) {
      Logger.warn(TAG, error)
      dispatch(showError(ErrorMessages.GET_BANK_ACCOUNTS_FAIL))
      return
    }
  }, [newPublicToken])

  function getBankDisplay(bank: BankAccount) {
    // Todo: Consider adding a default placeholder image for banks without a logo available
    const bankLogoSrc = bank.institutionLogo ? `data:image/png;base64,${bank.institutionLogo}` : ''
    return (
      <View key={bank.id} style={styles.accountContainer} testID="accountContainer">
        <View style={styles.row}>
          <View style={styles.bankImgContainer}>
            <Image
              style={styles.bankImg}
              source={{ uri: bankLogoSrc }}
              testID={`BankLogoImg-${bank.id}`}
            />
          </View>
          <View style={styles.accountLabels}>
            <Text style={styles.bankName}>{`${
              bank.institutionName
            } (${bank.accountNumberTruncated.slice(-8)})`}</Text>
          </View>
        </View>
        <View style={styles.rightSide}>
          <BorderlessButton
            testID={`TripleDot${bank.id}`}
            onPress={() => {
              setIsOptionsVisible(true)
              setSelectedBankId(bank.id)
            }}
          >
            <View style={styles.tripleDots}>
              <TripleDotVertical />
            </View>
          </BorderlessButton>
        </View>
      </View>
    )
  }

  async function deleteBankAccount() {
    setIsOptionsVisible(false)
    ValoraAnalytics.track(CICOEvents.delete_bank_account, {
      id: selectedBankId,
    })
    try {
      await deleteFinclusiveBankAccount({
        ...verifyWalletAddress({ walletAddress }),
        id: selectedBankId,
      })
      await bankAccounts.execute()
    } catch (error) {
      Logger.warn(TAG, error)
      dispatch(showError(ErrorMessages.DELETE_BANK_ACCOUNT_FAIL))
    }
  }

  return (
    <ScrollView style={styles.scrollContainer}>
      {bankAccounts?.loading && (
        <ActivityIndicator size="large" color={colors.gray2} testID="Loader" />
      )}
      {bankAccounts?.result?.map(getBankDisplay)}
      {!bankAccounts?.loading && (
        <View style={styles.addAccountContainer}>
          <BorderlessButton
            testID="AddAccount"
            onPress={async () => {
              ValoraAnalytics.track(CICOEvents.add_bank_account_start)
              await openPlaid({
                ...plaidParams,
                onSuccess: ({ publicToken }) => {
                  navigate(Screens.SyncBankAccountScreen, {
                    publicToken,
                  })
                },
                onExit: ({ error }) => {
                  if (error) {
                    navigate(Screens.LinkBankAccountErrorScreen, {
                      error,
                    })
                  }
                },
              })
            }}
          >
            <View style={styles.row}>
              <View style={styles.plusIconContainer}>
                <PlusIcon />
              </View>
              <View style={styles.accountLabels}>
                <Text style={styles.bankName}>{t('bankAccountsScreen.add')}</Text>
              </View>
            </View>
          </BorderlessButton>
        </View>
      )}
      <OptionsChooser
        isVisible={isOptionsVisible}
        options={[t('bankAccountsScreen.delete')]}
        includeCancelButton={true}
        isLastOptionDestructive={true}
        onOptionChosen={deleteBankAccount}
        onCancel={() => setIsOptionsVisible(false)}
      />
    </ScrollView>
  )
}

BankAccounts.navigationOptions = {
  ...emptyHeader,
}

const styles = StyleSheet.create({
  scrollContainer: {
    flex: 1,
    flexDirection: 'column',
    paddingHorizontal: variables.contentPadding,
    marginTop: 16,
  },
  accountContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    flex: 1,
    height: 72,
    borderBottomWidth: 1,
    borderColor: colors.gray2,
  },
  addAccountContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    flex: 1,
    height: 72,
    marginTop: 12,
  },
  accountLabels: {
    flexDirection: 'column',
  },
  rightSide: {
    flexDirection: 'column',
    alignItems: 'flex-end',
    justifyContent: 'center',
  },
  tripleDots: {
    padding: 10,
  },
  header: {
    flexDirection: 'column',
    alignItems: 'center',
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  bankName: {
    fontSize: 16,
  },
  bankImg: {
    width: 20,
    height: 20,
    borderRadius: 10,
    backgroundColor: colors.gray2,
  },
  bankImgContainer: {
    width: 40,
    height: 40,
    borderRadius: 20,
    marginRight: 12,
    padding: 10,
    backgroundColor: colors.gray2,
  },
  plusIconContainer: {
    width: 40,
    height: 40,
    borderRadius: 20,
    marginRight: 12,
    padding: 14,
    backgroundColor: colors.gray2,
  },
  backButton: {
    paddingLeft: 20,
  },
})

export default BankAccounts
