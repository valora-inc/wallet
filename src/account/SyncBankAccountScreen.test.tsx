import { render, waitFor } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { Actions } from 'src/account/actions'
import SyncBankAccountScreen from 'src/account/SyncBankAccountScreen'
import { createFinclusiveBankAccount, exchangePlaidAccessToken } from 'src/in-house-liquidity'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { createMockStore, getMockStackScreenProps } from 'test/utils'

const mockPublicToken = 'foo'
const mockAccessToken = 'bar'
const mockError = new Error('some error')

const mockProps = getMockStackScreenProps(Screens.SyncBankAccountScreen, {
  publicToken: mockPublicToken,
})

jest.mock('src/in-house-liquidity', () => ({
  ...(jest.requireActual('src/in-house-liquidity') as any),
  createFinclusiveBankAccount: jest.fn(() => Promise.resolve()),
  exchangePlaidAccessToken: jest.fn(() => Promise.resolve(mockAccessToken)),
}))

jest.mock('src/navigator/NavigationService', () => ({
  __esModule: true,
  namedExport: jest.fn(),
  default: jest.fn(),
  navigate: jest.fn(),
}))

describe('SyncBankAccountScreen', () => {
  const mockWalletAddress = '0x123'
  const store = createMockStore({
    web3: {
      account: mockWalletAddress,
    },
  })

  beforeEach(() => {
    store.dispatch = jest.fn()
    jest.useRealTimers()
    jest.clearAllMocks()
  })

  it('calls createFinclusiveBankAccount and exchangePlaidAccessToken', async () => {
    const { toJSON } = render(
      <Provider store={store}>
        <SyncBankAccountScreen {...mockProps} />
      </Provider>
    )
    expect(toJSON()).toMatchSnapshot()
    await waitFor(() => {
      expect(exchangePlaidAccessToken).toHaveBeenCalledWith({
        walletAddress: mockWalletAddress,
        publicToken: mockPublicToken,
      })
      expect(createFinclusiveBankAccount).toHaveBeenCalledWith({
        walletAddress: mockWalletAddress,
        plaidAccessToken: mockAccessToken,
      })
    })
    expect(store.dispatch).toHaveBeenCalledWith({ type: Actions.SET_HAS_LINKED_BANK_ACCOUNT })
  })

  it('directs to error page when token exchange fails', async () => {
    //@ts-ignore .
    exchangePlaidAccessToken.mockRejectedValue(mockError)

    render(
      <Provider store={store}>
        <SyncBankAccountScreen {...mockProps} />
      </Provider>
    )
    await waitFor(() => {
      expect(navigate).toHaveBeenCalledWith(Screens.LinkBankAccountErrorScreen, {
        error: mockError,
      })
    })
  })

  it('directs to error page when create finclusive bank account fails', async () => {
    //@ts-ignore .
    createFinclusiveBankAccount.mockRejectedValue(mockError)

    render(
      <Provider store={store}>
        <SyncBankAccountScreen {...mockProps} />
      </Provider>
    )
    await waitFor(() => {
      expect(navigate).toHaveBeenCalledWith(Screens.LinkBankAccountErrorScreen, {
        error: mockError,
      })
    })
  })
})
