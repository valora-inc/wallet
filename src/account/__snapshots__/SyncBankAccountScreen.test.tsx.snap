// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncBankAccountScreen calls createFinclusiveBankAccount and exchangePlaidAccessToken 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    color="#42D689"
    size="small"
  />
  <Text
    style={
      Object {
        "color": "#1AB775",
        "fontFamily": "Inter-Regular",
        "fontSize": 14,
        "lineHeight": 18,
        "marginTop": 20,
      }
    }
  >
    syncingBankAccount
  </Text>
  <Text
    style={
      Object {
        "color": "#81868B",
        "fontFamily": "Inter-Regular",
        "fontSize": 14,
        "lineHeight": 18,
        "marginTop": 20,
      }
    }
  >
    keepAppOpen
  </Text>
</View>
`;
